src/Illuminate/Bus/ChainedBatch.php

```diff
@@ -0,0 +1,87 @@
+<?php
+
+namespace Illuminate\Bus;
+
+use Illuminate\Contracts\Container\Container;
+use Illuminate\Contracts\Queue\ShouldQueue;
+use Illuminate\Foundation\Bus\Dispatchable;
+use Illuminate\Queue\InteractsWithQueue;
+use Illuminate\Support\Collection;
+
+class ChainedBatch implements ShouldQueue
+{
+    use Batchable, Dispatchable, InteractsWithQueue, Queueable;
+
+    public Collection $jobs;
+
+    public array $options;
+
+    public string $name;
+
+    public function __construct(PendingBatch $batch)
+    {
+        $this->jobs = static::prepareNestedBatches($batch->jobs);
+        $this->options = $batch->options;
+        $this->name = $batch->name;
+    }
+
+    public function handle(Container $container)
+    {
+        $batch = new PendingBatch($container, $this->jobs);
+        $batch->name = $this->name;
+        $batch->options = $this->options;
+
+        $this->hijackChain($batch);
+
+        if ($this->queue) {
+            $batch->onQueue($this->queue);
+        }
+
+        if ($this->connection) {
+            $batch->onConnection($this->connection);
+        }
+
+        foreach ($this->chainCatchCallbacks ?? [] as $cb) {
+            $batch->catch($cb);
+        }
+
+        $batch->dispatch();
+    }
+
+    protected function hijackChain(PendingBatch $batch)
```

Comment:
I'm having trouble understanding what this function aims to do.

Reply:
By default the next job in the chain would run as soon as soon as the job that was put on the chain is done. But in this case we want it to run only once the batch is finished. 
So we need to grab what's `next` on the job chain and get it to run as a `then` callback of the entire batch.